numba/core/untyped_passes.py

```diff
@@ -1680,6 +1681,103 @@ def run_pass(self, state):
         return mutated
 
 
+@register_pass(mutates_CFG=True, analysis_only=False)
+class DeadLoopElimination(FunctionPass):
+    _name = "dead_loop_elimination"
+
+    def __init__(self):
+        FunctionPass.__init__(self)
+        self.dead_vars = []
+
+    def is_arg_constsized(self, state, arg):
+        # True if len(type(arg)) == 0
+        t = state.typemap.get(arg.name)
+        return isinstance(t, types.ConstSized) and len(t) == 0
+
+    def match_iter(self, state, getiter):
+        # call_fn  := Var(...)
+        # call_ret := call_expr(func=call_fn, args=[...])
+        #        _ := getiter(value=call_ret)
+        func_ir = state.func_ir
+        call_expr = get_definition(func_ir, getiter.value)
+        if not (isinstance(call_expr, ir.Expr) and call_expr.op == 'call'):
+            return False
+        call_fn = get_definition(func_ir, call_expr.func)
+
+        if call_fn.value not in (iter, zip):
+            return False
+
+        for arg in call_expr.args:
+            # only one arg?
+            if self.is_arg_constsized(state, arg):
+                # store the call expr to be removed later
+                self.dead_vars.append(func_ir.get_assignee(call_expr))
+                return True
+        return False
+
+    def find_dead_loops(self, state):
+        dead_loops = list()
+        cfg = compute_cfg_from_blocks(state.func_ir.blocks)
+        for loop in cfg.loops().values():
+            header = state.func_ir.blocks[loop.header]
+            for expr in header.find_exprs('iternext'):
+                expr = guard(get_definition, state.func_ir, expr.value.name)
+                if expr:
+                    if isinstance(expr, ir.Expr) and expr.op == 'getiter':
+                        # expr := getiter(value=Var)
+                        if self.is_arg_constsized(state, expr.value) or \
+                           self.match_iter(state, expr):
+                            dead_loops.append(loop)
+        return dead_loops
+
+    def remove_dead_loops(self, state, dead_loops):
+        for loop in dead_loops:
+            # change branch into a direct jump to exit
+            header = state.func_ir.blocks[loop.header]
+            term = header.terminator
+            header.insert_before_terminator(ir.Jump(term.falsebr, term.loc))
+            header.remove(term)
+
+            # iterate over header and remove "pair_first", since
+            # dead_code_elimination won't remove it
+            for inst in header.find_insts(ir.Assign):
+                expr = inst.value
+                if isinstance(expr, ir.Expr) and expr.op == 'pair_first':
+                    header.remove(inst)
```

Comment:
`func_ir._definitions` will contain reference to variables that do not exist following this and the `header.remove` above?

Reply:
I'll reset the definitions once the pass execute

Reply:
@stuartarchibald, can you check if what I did seems correct?

Reply:
Thanks, the code added to rebuild looks approximately as I'd expect it to.